### PR TITLE
[FIX] test_invoice_refund

### PR DESCRIPTION
**Pull request by rven**
Trying to trigger the OCA runbot

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Why this?

**Comment by rven** [comment]
Latest version of Odoo gives following error:

2019-07-15 12:24:20,457 18065 ERROR odoo_test odoo.addons.account_payment_partner.tests.test_account_payment_partner: FAIL: test_invoice_refund (odoo.addons.account_payment_partner.tests.test_account_payment_partner.TestAccountPaymentPartner) 
2019-07-15 12:24:20,457 18065 ERROR odoo_test odoo.addons.account_payment_partner.tests.test_account_payment_partner: Traceback (most recent call last): 
2019-07-15 12:24:20,457 18065 ERROR odoo_test odoo.addons.account_payment_partner.tests.test_account_payment_partner: `   File "/opt/odoo/buildout/parts/oca/bank-payment/account_payment_partner/tests/test_account_payment_partner.py", line 310, in test_invoice_refund 
2019-07-15 12:24:20,457 18065 ERROR odoo_test odoo.addons.account_payment_partner.tests.test_account_payment_partner: `     invoice.partner_bank_id) 
2019-07-15 12:24:20,458 18065 ERROR odoo_test odoo.addons.account_payment_partner.tests.test_account_payment_partner: ` AssertionError: res.partner.bank(38,) != res.partner.bank() 
2019-07-15 12:24:20,458 18065 INFO odoo_test odoo.addons.account_payment_partner.tests.test_account_payment_partner: Ran 12 tests in 6.216s 
2019-07-15 12:24:20,458 18065 ERROR odoo_test odoo.addons.account_payment_partner.tests.test_account_payment_partner: FAILED 
2019-07-15 12:24:20,458 18065 INFO odoo_test odoo.addons.account_payment_partner.tests.test_account_payment_partner:  (failures=1) 
2019-07-15 12:24:20,458 18065 INFO odoo_test odoo.modules.module: odoo.addons.account_payment_partner.tests.test_account_payment_partner tested in 6.22s, 4085 queries 
2019-07-15 12:24:20,458 18065 ERROR odoo_test odoo.modules.module: Module account_payment_partner: 1 failures, 0 errors

**Comment by rven** [comment]
image

**Comment by pedrobaeza** [comment]
OK, then you should fix the problem instead of only adding you as contributor...

File: account_payment_partner/readme/CONTRIBUTORS.rst [hunk 1]
```diff
@@ -9,3 +9,5 @@
 
   * Pedro M. Baeza <pedro.baeza@tecnativa.com>
   * Carlos Dauden <carlos.dauden@tecnativa.com>
+* `DynApps <https://www.dynapps.be>`_:
```
Comment (pedrobaeza) [inline account_payment_partner/readme/CONTRIBUTORS.rst:12]: You must leave an empty line before and after for proper RST syntax

**Comment by pedrobaeza** [comment]
Aren't you going to fix it?

**Comment by rven** [comment]
Yes I am

**Comment by pedrobaeza** [comment]
OK, thanks :smiley:

File: account_payment_partner/README.rst [hunk 2]
```diff
@@ -111,6 +111,9 @@ Contributors
 
   * Pedro M. Baeza <pedro.baeza@tecnativa.com>
   * Carlos Dauden <carlos.dauden@tecnativa.com>
+* `DynApps <https://www.dynapps.be>`_:
```
Comment (pedrobaeza) [inline account_payment_partner/README.rst:114]: Leave an empty line before

Comment (rven) [reply account_payment_partner/README.rst:114]: image
Before is not needed?

Comment (pedrobaeza) [reply account_payment_partner/README.rst:114]: OK, I see... I though it was, sorry.

**Review by pedrobaeza — APPROVED** [review]

File: account_payment_partner/README.rst (hunk 2 above)
Comment (rven) [reply account_payment_partner/README.rst:114]: It has no influence on the result...

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Hey, wait, I'm seeing that there's specific code that tries to avoid this behavior:

https://github.com/OCA/bank-payment/blob/7c5cb9638b4139fe25e1df605f1b4f90345aab2a/account_payment_partner/models/account_invoice.py#L95-L109

so we shouldn't reset the test, but fixing the behavior

**Comment by pedrobaeza** [comment]
Or should we leave the same bank account? @sbidoul @alexis-via can you check this new upstream behavior?

**Comment by rven** [comment]
Isn't it ok.

```python
    def _get_default_bank_id(self, type, company_id):
        """When setting the default bank account, we have two cases:
         in the case of 'out_invoice', 'in_refund', we want the bank account of the company.
         otherwise, we want the bank account of the partner.
        """
        if type not in ('out_invoice', 'in_refund') or not company_id:
            return False
        return self._get_partner_bank_id(company_id)
```

**Comment by pedrobaeza** [comment]
Any news, @sbidoul @alexis-via ?

**Comment by sbidoul** [comment]
So if I understand this thread correctly, Odoo changed behaviour when creating refunds from invoices?

I would say the test is correct (i.e. preserve the bank account from the invoice in the refund), for the same reason it makes sense to preserve the invoice payment mode in the refund.

So at first sight, I'd say:
- we need to fix the behaviour and preserve the test
- raise the issue at Odoo?
- see what it does for customer invoices (in case of direct debit): IMO the bank account must be preserved in refunds too

**Comment by pedrobaeza** [comment]
Yeah, in a nutshell, Odoo considers that on refund, it should be changed, and thus, the error in test. Personally, I don't care too much about bank account indicated there, as it's almost ignored in all places by bank-payment/l10n-spain suite. But it seems others like @alexis-via relies a lot on that data.

We must discard option 2, as Odoo has made this change because they think the contrary, so the question is to merge this PR, assuming the new behavior is correct, changing the test, or overwrite behavior for keeping old one, and don't change test.

**Comment by sbidoul** [comment]
I'm not sure Odoo made this change intentionally. As a default value, their change can make sense, but when creating a refund from an invoice the default may not apply the same way.

In any case I vote for keeping the test as is and fix the behaviour, to be consistent with the copy of the payment mode.

**Comment by pedrobaeza** [comment]
If you see the source commit, it was an OPW fixed that way.

**Comment by pedrobaeza** [comment]
Let's fix the branch for now merging this and if we change the behavior later, we can revert back to the original test.

**Pull request merged**